core_java/grep/src/main/java/ca/jrvs/apps/grep/JavaGrepImp.java

```diff
@@ -0,0 +1,173 @@
+package ca.jrvs.apps.grep;
+
+import com.sun.org.slf4j.internal.Logger;
+import com.sun.org.slf4j.internal.LoggerFactory;
+import java.io.File;
+import java.io.FileNotFoundException;
+import java.io.FileWriter;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Scanner;
+
+public class JavaGrepImp implements JavaGrep {
+  final Logger logger = LoggerFactory.getLogger(JavaGrep.class);
+  private String regex;
+  private String rootPath;
+  private String outFile;
+
+  public static void main(String[] args) {
+    if (args.length != 3 ) {
+      throw new IllegalArgumentException("USAGE: Java regex rootPath outFile");
+    }
+    JavaGrepImp JavaGrepImp = new JavaGrepImp();
+    JavaGrepImp.setRegex(args[0]);
+    JavaGrepImp.setRootPath(args[1]);
+    JavaGrepImp.setOutFile(args[2]);
+
+    try {
+      JavaGrepImp.process();
+    }catch (Exception ex){
+      JavaGrepImp.logger.error(ex.getMessage());
+    }
+  }
+  /**
+   * Top level search workflow
+   *
+   * @throws IOException
+   */
+  @Override
+  public void process() throws IOException {
+    // initialize a List of string to store all the matched lines
+    List<String> matchedLines = new ArrayList<>();
+    try {
+      //initialize and store all the files
+      for (File file : listFiles(getRootPath())) {
+        // for each file go through parse the file and read each line
+        for (String line : readLines(file)) {
+          // for each line if that line matches the pattern store it in matchedLines
+          if (containsPattern(line)) {
+            matchedLines.add(line);
+          }
+        }
+      }
+    } catch(NullPointerException e){
+      e.printStackTrace();
+    }
+    writeToFile(matchedLines);
+  }
+
+  /**
+   * Traverse a given directory and return all files
+   *
+   * @param rootDir input directory
+   * @return files under rootDir
+   */
+  @Override
+  public List<File> listFiles(String rootDir) {
+    //initialize return list
+    List<File> res = new ArrayList<>();
+    //Open rootDir and get the contents of rootDir
+    File directory = new File(rootDir);
+    File[] directoryContents = directory.listFiles();
+    //if Directory is empty return null
```

Comment:
remove all inline comments as any developer could understand by reading the code